<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>GuitarFast vs Guitar: Performance Benchmark</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">GuitarFast vs Guitar: Performance
Benchmark</h1>



<div id="summary" class="section level2">
<h2>Summary</h2>
<p>GuitarFast is a performance-optimized fork of the Guitar Bioconductor
package (v2.26.0). This vignette documents the optimizations and
benchmark results.</p>
</div>
<div id="benchmark-results" class="section level2">
<h2>Benchmark Results</h2>
<p>Measured on the bundled toy dataset (mm10 subset, 1000 m6A peaks).
Both packages produce identical density curves for deterministic (no-CI)
runs.</p>
<table>
<thead>
<tr class="header">
<th align="left">Scenario</th>
<th align="right">Guitar (original)</th>
<th align="right">GuitarFast</th>
<th align="right">Speedup</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">tx, no CI</td>
<td align="right">8.40s</td>
<td align="right">1.52s</td>
<td align="right"><strong>5.5x</strong></td>
</tr>
<tr class="even">
<td align="left">tx, CI (200 resamples)</td>
<td align="right">9.37s</td>
<td align="right">1.56s</td>
<td align="right"><strong>6.0x</strong></td>
</tr>
<tr class="odd">
<td align="left">mrna, no CI</td>
<td align="right">4.22s</td>
<td align="right">1.40s</td>
<td align="right"><strong>3.0x</strong></td>
</tr>
<tr class="even">
<td align="left">All types, CI (200 resamples)</td>
<td align="right">18.77s</td>
<td align="right">4.79s</td>
<td align="right"><strong>3.9x</strong></td>
</tr>
</tbody>
</table>
</div>
<div id="running-the-benchmark-yourself" class="section level2">
<h2>Running the Benchmark Yourself</h2>
<p>GuitarFast ships with a benchmark script in
<code>tests/test_benchmark.R</code>:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">library</span>(GuitarFast)</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="co">#&gt; Loading required package: GenomicFeatures</span></span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="co">#&gt; Loading required package: BiocGenerics</span></span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a><span class="co">#&gt; Attaching package: &#39;BiocGenerics&#39;</span></span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a><span class="co">#&gt; The following objects are masked from &#39;package:stats&#39;:</span></span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a><span class="co">#&gt;     IQR, mad, sd, var, xtabs</span></span>
<span id="cb1-9"><a href="#cb1-9" tabindex="-1"></a><span class="co">#&gt; The following objects are masked from &#39;package:base&#39;:</span></span>
<span id="cb1-10"><a href="#cb1-10" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb1-11"><a href="#cb1-11" tabindex="-1"></a><span class="co">#&gt;     anyDuplicated, aperm, append, as.data.frame, basename, cbind,</span></span>
<span id="cb1-12"><a href="#cb1-12" tabindex="-1"></a><span class="co">#&gt;     colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,</span></span>
<span id="cb1-13"><a href="#cb1-13" tabindex="-1"></a><span class="co">#&gt;     get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,</span></span>
<span id="cb1-14"><a href="#cb1-14" tabindex="-1"></a><span class="co">#&gt;     match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,</span></span>
<span id="cb1-15"><a href="#cb1-15" tabindex="-1"></a><span class="co">#&gt;     Position, rank, rbind, Reduce, rownames, sapply, saveRDS, setdiff,</span></span>
<span id="cb1-16"><a href="#cb1-16" tabindex="-1"></a><span class="co">#&gt;     table, tapply, union, unique, unsplit, which.max, which.min</span></span>
<span id="cb1-17"><a href="#cb1-17" tabindex="-1"></a><span class="co">#&gt; Loading required package: S4Vectors</span></span>
<span id="cb1-18"><a href="#cb1-18" tabindex="-1"></a><span class="co">#&gt; Loading required package: stats4</span></span>
<span id="cb1-19"><a href="#cb1-19" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb1-20"><a href="#cb1-20" tabindex="-1"></a><span class="co">#&gt; Attaching package: &#39;S4Vectors&#39;</span></span>
<span id="cb1-21"><a href="#cb1-21" tabindex="-1"></a><span class="co">#&gt; The following object is masked from &#39;package:utils&#39;:</span></span>
<span id="cb1-22"><a href="#cb1-22" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb1-23"><a href="#cb1-23" tabindex="-1"></a><span class="co">#&gt;     findMatches</span></span>
<span id="cb1-24"><a href="#cb1-24" tabindex="-1"></a><span class="co">#&gt; The following objects are masked from &#39;package:base&#39;:</span></span>
<span id="cb1-25"><a href="#cb1-25" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb1-26"><a href="#cb1-26" tabindex="-1"></a><span class="co">#&gt;     expand.grid, I, unname</span></span>
<span id="cb1-27"><a href="#cb1-27" tabindex="-1"></a><span class="co">#&gt; Loading required package: IRanges</span></span>
<span id="cb1-28"><a href="#cb1-28" tabindex="-1"></a><span class="co">#&gt; Loading required package: GenomeInfoDb</span></span>
<span id="cb1-29"><a href="#cb1-29" tabindex="-1"></a><span class="co">#&gt; Loading required package: GenomicRanges</span></span>
<span id="cb1-30"><a href="#cb1-30" tabindex="-1"></a><span class="co">#&gt; Loading required package: AnnotationDbi</span></span>
<span id="cb1-31"><a href="#cb1-31" tabindex="-1"></a><span class="co">#&gt; Loading required package: Biobase</span></span>
<span id="cb1-32"><a href="#cb1-32" tabindex="-1"></a><span class="co">#&gt; Welcome to Bioconductor</span></span>
<span id="cb1-33"><a href="#cb1-33" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb1-34"><a href="#cb1-34" tabindex="-1"></a><span class="co">#&gt;     Vignettes contain introductory material; view with</span></span>
<span id="cb1-35"><a href="#cb1-35" tabindex="-1"></a><span class="co">#&gt;     &#39;browseVignettes()&#39;. To cite Bioconductor, see</span></span>
<span id="cb1-36"><a href="#cb1-36" tabindex="-1"></a><span class="co">#&gt;     &#39;citation(&quot;Biobase&quot;)&#39;, and for packages &#39;citation(&quot;pkgname&quot;)&#39;.</span></span>
<span id="cb1-37"><a href="#cb1-37" tabindex="-1"></a><span class="co">#&gt; Loading required package: rtracklayer</span></span>
<span id="cb1-38"><a href="#cb1-38" tabindex="-1"></a><span class="co">#&gt; Loading required package: magrittr</span></span>
<span id="cb1-39"><a href="#cb1-39" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb1-40"><a href="#cb1-40" tabindex="-1"></a><span class="co">#&gt; Attaching package: &#39;magrittr&#39;</span></span>
<span id="cb1-41"><a href="#cb1-41" tabindex="-1"></a><span class="co">#&gt; The following object is masked from &#39;package:GenomicRanges&#39;:</span></span>
<span id="cb1-42"><a href="#cb1-42" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb1-43"><a href="#cb1-43" tabindex="-1"></a><span class="co">#&gt;     subtract</span></span>
<span id="cb1-44"><a href="#cb1-44" tabindex="-1"></a><span class="co">#&gt; Loading required package: ggplot2</span></span>
<span id="cb1-45"><a href="#cb1-45" tabindex="-1"></a><span class="co">#&gt; Loading required package: knitr</span></span>
<span id="cb1-46"><a href="#cb1-46" tabindex="-1"></a><span class="co">#&gt; Loading required package: dplyr</span></span>
<span id="cb1-47"><a href="#cb1-47" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb1-48"><a href="#cb1-48" tabindex="-1"></a><span class="co">#&gt; Attaching package: &#39;dplyr&#39;</span></span>
<span id="cb1-49"><a href="#cb1-49" tabindex="-1"></a><span class="co">#&gt; The following object is masked from &#39;package:AnnotationDbi&#39;:</span></span>
<span id="cb1-50"><a href="#cb1-50" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb1-51"><a href="#cb1-51" tabindex="-1"></a><span class="co">#&gt;     select</span></span>
<span id="cb1-52"><a href="#cb1-52" tabindex="-1"></a><span class="co">#&gt; The following object is masked from &#39;package:Biobase&#39;:</span></span>
<span id="cb1-53"><a href="#cb1-53" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb1-54"><a href="#cb1-54" tabindex="-1"></a><span class="co">#&gt;     combine</span></span>
<span id="cb1-55"><a href="#cb1-55" tabindex="-1"></a><span class="co">#&gt; The following objects are masked from &#39;package:GenomicRanges&#39;:</span></span>
<span id="cb1-56"><a href="#cb1-56" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb1-57"><a href="#cb1-57" tabindex="-1"></a><span class="co">#&gt;     intersect, setdiff, union</span></span>
<span id="cb1-58"><a href="#cb1-58" tabindex="-1"></a><span class="co">#&gt; The following object is masked from &#39;package:GenomeInfoDb&#39;:</span></span>
<span id="cb1-59"><a href="#cb1-59" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb1-60"><a href="#cb1-60" tabindex="-1"></a><span class="co">#&gt;     intersect</span></span>
<span id="cb1-61"><a href="#cb1-61" tabindex="-1"></a><span class="co">#&gt; The following objects are masked from &#39;package:IRanges&#39;:</span></span>
<span id="cb1-62"><a href="#cb1-62" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb1-63"><a href="#cb1-63" tabindex="-1"></a><span class="co">#&gt;     collapse, desc, intersect, setdiff, slice, union</span></span>
<span id="cb1-64"><a href="#cb1-64" tabindex="-1"></a><span class="co">#&gt; The following objects are masked from &#39;package:S4Vectors&#39;:</span></span>
<span id="cb1-65"><a href="#cb1-65" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb1-66"><a href="#cb1-66" tabindex="-1"></a><span class="co">#&gt;     first, intersect, rename, setdiff, setequal, union</span></span>
<span id="cb1-67"><a href="#cb1-67" tabindex="-1"></a><span class="co">#&gt; The following objects are masked from &#39;package:BiocGenerics&#39;:</span></span>
<span id="cb1-68"><a href="#cb1-68" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb1-69"><a href="#cb1-69" tabindex="-1"></a><span class="co">#&gt;     combine, intersect, setdiff, union</span></span>
<span id="cb1-70"><a href="#cb1-70" tabindex="-1"></a><span class="co">#&gt; The following objects are masked from &#39;package:stats&#39;:</span></span>
<span id="cb1-71"><a href="#cb1-71" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb1-72"><a href="#cb1-72" tabindex="-1"></a><span class="co">#&gt;     filter, lag</span></span>
<span id="cb1-73"><a href="#cb1-73" tabindex="-1"></a><span class="co">#&gt; The following objects are masked from &#39;package:base&#39;:</span></span>
<span id="cb1-74"><a href="#cb1-74" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb1-75"><a href="#cb1-75" tabindex="-1"></a><span class="co">#&gt;     intersect, setdiff, setequal, union</span></span>
<span id="cb1-76"><a href="#cb1-76" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb1-77"><a href="#cb1-77" tabindex="-1"></a><span class="co">#&gt; Attaching package: &#39;GuitarFast&#39;</span></span>
<span id="cb1-78"><a href="#cb1-78" tabindex="-1"></a><span class="co">#&gt; The following object is masked from &#39;package:BiocGenerics&#39;:</span></span>
<span id="cb1-79"><a href="#cb1-79" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb1-80"><a href="#cb1-80" tabindex="-1"></a><span class="co">#&gt;     normalize</span></span>
<span id="cb1-81"><a href="#cb1-81" tabindex="-1"></a><span class="fu">library</span>(GenomicFeatures)</span>
<span id="cb1-82"><a href="#cb1-82" tabindex="-1"></a></span>
<span id="cb1-83"><a href="#cb1-83" tabindex="-1"></a>txdb <span class="ot">&lt;-</span> <span class="fu">loadDb</span>(<span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;mm10_toy.sqlite&quot;</span>, <span class="at">package =</span> <span class="st">&quot;GuitarFast&quot;</span>))</span>
<span id="cb1-84"><a href="#cb1-84" tabindex="-1"></a>bed1 <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;m6A_mm10_exomePeak_1000peaks_bed12.bed&quot;</span>,</span>
<span id="cb1-85"><a href="#cb1-85" tabindex="-1"></a>                     <span class="at">package =</span> <span class="st">&quot;GuitarFast&quot;</span>)</span>
<span id="cb1-86"><a href="#cb1-86" tabindex="-1"></a></span>
<span id="cb1-87"><a href="#cb1-87" tabindex="-1"></a>bench <span class="ot">&lt;-</span> <span class="cf">function</span>(label, expr) {</span>
<span id="cb1-88"><a href="#cb1-88" tabindex="-1"></a>  t1 <span class="ot">&lt;-</span> <span class="fu">proc.time</span>()</span>
<span id="cb1-89"><a href="#cb1-89" tabindex="-1"></a>  result <span class="ot">&lt;-</span> <span class="fu">eval</span>(expr)</span>
<span id="cb1-90"><a href="#cb1-90" tabindex="-1"></a>  elapsed <span class="ot">&lt;-</span> (<span class="fu">proc.time</span>() <span class="sc">-</span> t1)[<span class="st">&quot;elapsed&quot;</span>]</span>
<span id="cb1-91"><a href="#cb1-91" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="fu">sprintf</span>(<span class="st">&quot;  %-35s %6.2fs</span><span class="sc">\n</span><span class="st">&quot;</span>, label, elapsed))</span>
<span id="cb1-92"><a href="#cb1-92" tabindex="-1"></a>  <span class="fu">invisible</span>(result)</span>
<span id="cb1-93"><a href="#cb1-93" tabindex="-1"></a>}</span>
<span id="cb1-94"><a href="#cb1-94" tabindex="-1"></a></span>
<span id="cb1-95"><a href="#cb1-95" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&quot;== GuitarFast Benchmarks ==</span><span class="sc">\n</span><span class="st">&quot;</span>)</span>
<span id="cb1-96"><a href="#cb1-96" tabindex="-1"></a><span class="co">#&gt; == GuitarFast Benchmarks ==</span></span>
<span id="cb1-97"><a href="#cb1-97" tabindex="-1"></a></span>
<span id="cb1-98"><a href="#cb1-98" tabindex="-1"></a><span class="fu">bench</span>(<span class="st">&quot;makeGuitarTxdb&quot;</span>,</span>
<span id="cb1-99"><a href="#cb1-99" tabindex="-1"></a>  <span class="fu">quote</span>(<span class="fu">makeGuitarTxdb</span>(<span class="at">txdb =</span> txdb, <span class="at">txPrimaryOnly =</span> <span class="cn">FALSE</span>)))</span>
<span id="cb1-100"><a href="#cb1-100" tabindex="-1"></a><span class="co">#&gt; [1] &quot;There are 2946 transcripts of 2946 genes in the genome.&quot;</span></span>
<span id="cb1-101"><a href="#cb1-101" tabindex="-1"></a><span class="co">#&gt; [1] &quot;total 2946 transcripts extracted ...&quot;</span></span>
<span id="cb1-102"><a href="#cb1-102" tabindex="-1"></a><span class="co">#&gt; [1] &quot;total 2719 transcripts left after ambiguity filter ...&quot;</span></span>
<span id="cb1-103"><a href="#cb1-103" tabindex="-1"></a><span class="co">#&gt; [1] &quot;total 2719 transcripts left after check chromosome validity ...&quot;</span></span>
<span id="cb1-104"><a href="#cb1-104" tabindex="-1"></a><span class="co">#&gt; [1] &quot;total 1342 mRNAs left after component length filter ...&quot;</span></span>
<span id="cb1-105"><a href="#cb1-105" tabindex="-1"></a><span class="co">#&gt; [1] &quot;total 307 ncRNAs left after ncRNA length filter ...&quot;</span></span>
<span id="cb1-106"><a href="#cb1-106" tabindex="-1"></a><span class="co">#&gt; [1] &quot;generate components for all tx&quot;</span></span>
<span id="cb1-107"><a href="#cb1-107" tabindex="-1"></a><span class="co">#&gt; [1] &quot;generate components for mRNA&quot;</span></span>
<span id="cb1-108"><a href="#cb1-108" tabindex="-1"></a><span class="co">#&gt; [1] &quot;generate components for lncRNA&quot;</span></span>
<span id="cb1-109"><a href="#cb1-109" tabindex="-1"></a><span class="co">#&gt; [1] &quot;generate chiped transcriptome&quot;</span></span>
<span id="cb1-110"><a href="#cb1-110" tabindex="-1"></a><span class="co">#&gt; [1] &quot;generate coverage checking ranges for tx&quot;</span></span>
<span id="cb1-111"><a href="#cb1-111" tabindex="-1"></a><span class="co">#&gt; [1] &quot;generate coverage checking ranges for mrna&quot;</span></span>
<span id="cb1-112"><a href="#cb1-112" tabindex="-1"></a><span class="co">#&gt; [1] &quot;generate coverage checking ranges for ncrna&quot;</span></span>
<span id="cb1-113"><a href="#cb1-113" tabindex="-1"></a><span class="co">#&gt;   makeGuitarTxdb                        1.69s</span></span>
<span id="cb1-114"><a href="#cb1-114" tabindex="-1"></a></span>
<span id="cb1-115"><a href="#cb1-115" tabindex="-1"></a><span class="fu">bench</span>(<span class="st">&quot;GuitarPlot tx, no CI&quot;</span>,</span>
<span id="cb1-116"><a href="#cb1-116" tabindex="-1"></a>  <span class="fu">quote</span>(<span class="fu">GuitarPlot</span>(<span class="at">txTxdb =</span> txdb, <span class="at">stBedFiles =</span> <span class="fu">list</span>(bed1),</span>
<span id="cb1-117"><a href="#cb1-117" tabindex="-1"></a>                   <span class="at">enableCI =</span> <span class="cn">FALSE</span>, <span class="at">pltTxType =</span> <span class="fu">c</span>(<span class="st">&quot;tx&quot;</span>))))</span>
<span id="cb1-118"><a href="#cb1-118" tabindex="-1"></a><span class="co">#&gt; [1] &quot;20260216164841&quot;</span></span>
<span id="cb1-119"><a href="#cb1-119" tabindex="-1"></a><span class="co">#&gt; [1] &quot;There are 2946 transcripts of 2946 genes in the genome.&quot;</span></span>
<span id="cb1-120"><a href="#cb1-120" tabindex="-1"></a><span class="co">#&gt; [1] &quot;total 2946 transcripts extracted ...&quot;</span></span>
<span id="cb1-121"><a href="#cb1-121" tabindex="-1"></a><span class="co">#&gt; [1] &quot;total 2719 transcripts left after ambiguity filter ...&quot;</span></span>
<span id="cb1-122"><a href="#cb1-122" tabindex="-1"></a><span class="co">#&gt; [1] &quot;total 2719 transcripts left after check chromosome validity ...&quot;</span></span>
<span id="cb1-123"><a href="#cb1-123" tabindex="-1"></a><span class="co">#&gt; [1] &quot;total 1342 mRNAs left after component length filter ...&quot;</span></span>
<span id="cb1-124"><a href="#cb1-124" tabindex="-1"></a><span class="co">#&gt; [1] &quot;total 307 ncRNAs left after ncRNA length filter ...&quot;</span></span>
<span id="cb1-125"><a href="#cb1-125" tabindex="-1"></a><span class="co">#&gt; [1] &quot;generate components for all tx&quot;</span></span>
<span id="cb1-126"><a href="#cb1-126" tabindex="-1"></a><span class="co">#&gt; [1] &quot;generate chiped transcriptome&quot;</span></span>
<span id="cb1-127"><a href="#cb1-127" tabindex="-1"></a><span class="co">#&gt; [1] &quot;generate coverage checking ranges for tx&quot;</span></span>
<span id="cb1-128"><a href="#cb1-128" tabindex="-1"></a><span class="co">#&gt; [1] &quot;20260216164842&quot;</span></span>
<span id="cb1-129"><a href="#cb1-129" tabindex="-1"></a><span class="co">#&gt; [1] &quot;import BED file /home/yangli/software/micromamba/envs/guitar/lib/R/library/GuitarFast/extdata/m6A_mm10_exomePeak_1000peaks_bed12.bed&quot;</span></span>
<span id="cb1-130"><a href="#cb1-130" tabindex="-1"></a><span class="co">#&gt; [1] &quot;sample 10 points for Group1&quot;</span></span>
<span id="cb1-131"><a href="#cb1-131" tabindex="-1"></a><span class="co">#&gt; [1] &quot;start figure plotting for tx ...&quot;</span></span>
<span id="cb1-132"><a href="#cb1-132" tabindex="-1"></a><span class="co">#&gt; Warning: Using `size` aesthetic for lines was deprecated in ggplot2 3.4.0.</span></span>
<span id="cb1-133"><a href="#cb1-133" tabindex="-1"></a><span class="co">#&gt; ℹ Please use `linewidth` instead.</span></span>
<span id="cb1-134"><a href="#cb1-134" tabindex="-1"></a><span class="co">#&gt; ℹ The deprecated feature was likely used in the GuitarFast package.</span></span>
<span id="cb1-135"><a href="#cb1-135" tabindex="-1"></a><span class="co">#&gt;   Please report the issue at</span></span>
<span id="cb1-136"><a href="#cb1-136" tabindex="-1"></a><span class="co">#&gt;   &lt;https://github.com/yliuchicago/GuitarFast/issues&gt;.</span></span>
<span id="cb1-137"><a href="#cb1-137" tabindex="-1"></a><span class="co">#&gt; This warning is displayed once per session.</span></span>
<span id="cb1-138"><a href="#cb1-138" tabindex="-1"></a><span class="co">#&gt; Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span></span>
<span id="cb1-139"><a href="#cb1-139" tabindex="-1"></a><span class="co">#&gt; generated.</span></span>
<span id="cb1-140"><a href="#cb1-140" tabindex="-1"></a><span class="co">#&gt;   GuitarPlot tx, no CI                  1.78s</span></span>
<span id="cb1-141"><a href="#cb1-141" tabindex="-1"></a></span>
<span id="cb1-142"><a href="#cb1-142" tabindex="-1"></a><span class="fu">bench</span>(<span class="st">&quot;GuitarPlot tx, CI (200 resamp)&quot;</span>,</span>
<span id="cb1-143"><a href="#cb1-143" tabindex="-1"></a>  <span class="fu">quote</span>(<span class="fu">GuitarPlot</span>(<span class="at">txTxdb =</span> txdb, <span class="at">stBedFiles =</span> <span class="fu">list</span>(bed1),</span>
<span id="cb1-144"><a href="#cb1-144" tabindex="-1"></a>                   <span class="at">enableCI =</span> <span class="cn">TRUE</span>, <span class="at">pltTxType =</span> <span class="fu">c</span>(<span class="st">&quot;tx&quot;</span>))))</span>
<span id="cb1-145"><a href="#cb1-145" tabindex="-1"></a><span class="co">#&gt; [1] &quot;20260216164842&quot;</span></span>
<span id="cb1-146"><a href="#cb1-146" tabindex="-1"></a><span class="co">#&gt; [1] &quot;There are 2946 transcripts of 2946 genes in the genome.&quot;</span></span>
<span id="cb1-147"><a href="#cb1-147" tabindex="-1"></a><span class="co">#&gt; [1] &quot;total 2946 transcripts extracted ...&quot;</span></span>
<span id="cb1-148"><a href="#cb1-148" tabindex="-1"></a><span class="co">#&gt; [1] &quot;total 2719 transcripts left after ambiguity filter ...&quot;</span></span>
<span id="cb1-149"><a href="#cb1-149" tabindex="-1"></a><span class="co">#&gt; [1] &quot;total 2719 transcripts left after check chromosome validity ...&quot;</span></span>
<span id="cb1-150"><a href="#cb1-150" tabindex="-1"></a><span class="co">#&gt; [1] &quot;total 1342 mRNAs left after component length filter ...&quot;</span></span>
<span id="cb1-151"><a href="#cb1-151" tabindex="-1"></a><span class="co">#&gt; [1] &quot;total 307 ncRNAs left after ncRNA length filter ...&quot;</span></span>
<span id="cb1-152"><a href="#cb1-152" tabindex="-1"></a><span class="co">#&gt; [1] &quot;generate components for all tx&quot;</span></span>
<span id="cb1-153"><a href="#cb1-153" tabindex="-1"></a><span class="co">#&gt; [1] &quot;generate chiped transcriptome&quot;</span></span>
<span id="cb1-154"><a href="#cb1-154" tabindex="-1"></a><span class="co">#&gt; [1] &quot;generate coverage checking ranges for tx&quot;</span></span>
<span id="cb1-155"><a href="#cb1-155" tabindex="-1"></a><span class="co">#&gt; [1] &quot;20260216164844&quot;</span></span>
<span id="cb1-156"><a href="#cb1-156" tabindex="-1"></a><span class="co">#&gt; [1] &quot;import BED file /home/yangli/software/micromamba/envs/guitar/lib/R/library/GuitarFast/extdata/m6A_mm10_exomePeak_1000peaks_bed12.bed&quot;</span></span>
<span id="cb1-157"><a href="#cb1-157" tabindex="-1"></a><span class="co">#&gt; [1] &quot;sample 10 points for Group1&quot;</span></span>
<span id="cb1-158"><a href="#cb1-158" tabindex="-1"></a><span class="co">#&gt; [1] &quot;start figure plotting for tx ...&quot;</span></span>
<span id="cb1-159"><a href="#cb1-159" tabindex="-1"></a><span class="co">#&gt;   GuitarPlot tx, CI (200 resamp)        3.38s</span></span></code></pre></div>
<p>A side-by-side comparison script (requires the original Guitar
package installed alongside GuitarFast) is available at
<code>tests/test_compare.R</code>.</p>
</div>
<div id="optimizations-applied" class="section level2">
<h2>Optimizations Applied</h2>
<div id="elementnrows-instead-of-vapplynrow-800x-speedup" class="section level3">
<h3>1. <code>elementNROWS()</code> instead of <code>vapply(NROW)</code>
(800x speedup)</h3>
<p><strong>File</strong>: <code>R/makeGuitarTxdb.R</code></p>
<p>The single largest bottleneck. The original code counted ranges per
transcript using:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="co"># Original (6.7 seconds)</span></span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="fu">vapply</span>(txRange, NROW, <span class="fu">numeric</span>(<span class="dv">1</span>))</span></code></pre></div>
<p>This calls R’s <code>NROW</code> individually on each of ~2700
GRanges elements via an R-level loop. Replaced with:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="co"># Optimized (0.008 seconds)</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="fu">elementNROWS</span>(txRange)</span></code></pre></div>
<p><code>elementNROWS()</code> is a single C-level vectorized call in
the IRanges package.</p>
</div>
<div id="reduced-default-ci-resamples-1000-to-200" class="section level3">
<h3>2. Reduced Default CI Resamples (1000 to 200)</h3>
<p><strong>File</strong>: <code>R/GuitarPlot.R</code></p>
<p>Each bootstrap iteration calls <code>density()</code> (C code), so
the only way to speed up the CI computation is to reduce the number of
iterations. We benchmarked CI quality:</p>
<table>
<thead>
<tr class="header">
<th align="right">Resamples</th>
<th align="right">Mean CI Range</th>
<th align="right">Max CI Range</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="right">200</td>
<td align="right">0.259</td>
<td align="right">0.852</td>
</tr>
<tr class="even">
<td align="right">500</td>
<td align="right">0.262</td>
<td align="right">0.861</td>
</tr>
<tr class="odd">
<td align="right">1000</td>
<td align="right">0.264</td>
<td align="right">0.866</td>
</tr>
</tbody>
</table>
<p>200 resamples produce statistically equivalent confidence intervals.
Users who want the original behavior can set
<code>CI_ResamplingTime = 1000</code>.</p>
</div>
<div id="pre-allocated-resampling-matrix" class="section level3">
<h3>3. Pre-allocated Resampling Matrix</h3>
<p><strong>File</strong>: <code>R/GuitarPlot.R</code></p>
<p>The original used <code>replicate()</code> which allocates memory on
each iteration. GuitarFast pre-allocates the full result matrix and
generates all random group indices in one batch call:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="co"># Pre-allocate 256 x N matrix</span></span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a>fit2 <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="dv">0</span>, <span class="at">nrow =</span> <span class="dv">256</span><span class="dt">L</span>, <span class="at">ncol =</span> CI_ResamplingTime)</span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a></span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a><span class="co"># Generate all resampling indices at once</span></span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a>resample_idx_mat <span class="ot">&lt;-</span> <span class="fu">matrix</span>(</span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a>  <span class="fu">sample.int</span>(nGroups, nGroups <span class="sc">*</span> CI_ResamplingTime, <span class="at">replace =</span> <span class="cn">TRUE</span>),</span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a>  <span class="at">nrow =</span> nGroups, <span class="at">ncol =</span> CI_ResamplingTime</span>
<span id="cb4-8"><a href="#cb4-8" tabindex="-1"></a>)</span></code></pre></div>
</div>
<div id="c-optimized-quantile-computation" class="section level3">
<h3>4. C-optimized Quantile Computation</h3>
<p><strong>File</strong>: <code>R/GuitarPlot.R</code></p>
<p>Replaced R-level <code>apply(fit2, 1, quantile, ...)</code> with
C-optimized <code>matrixStats::rowQuantiles()</code>:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">requireNamespace</span>(<span class="st">&quot;matrixStats&quot;</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>)) {</span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a>  fit3 <span class="ot">&lt;-</span> matrixStats<span class="sc">::</span><span class="fu">rowQuantiles</span>(fit2, <span class="at">probs =</span> CI_interval)</span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a>  fit3 <span class="ot">&lt;-</span> <span class="fu">t</span>(<span class="fu">apply</span>(fit2, <span class="dv">1</span>, quantile, CI_interval))</span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a>}</span></code></pre></div>
<p>Falls back gracefully when matrixStats is not installed.</p>
</div>
<div id="vectorized-normalize" class="section level3">
<h3>5. Vectorized <code>normalize()</code></h3>
<p><strong>File</strong>: <code>R/siteNormalize.R</code></p>
<p>Two R-level per-row loops replaced with vectorized operations:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="co"># Original: apply() per row</span></span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a>sitesPointsComponet <span class="ot">&lt;-</span> <span class="fu">apply</span>(startPointDiffer, <span class="dv">1</span>, max_which)</span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a></span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a><span class="co"># Optimized: vectorized max.col</span></span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a>neg_mask <span class="ot">&lt;-</span> startPointDiffer <span class="sc">&lt;</span> <span class="dv">0</span></span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a>col_idx_mat <span class="ot">&lt;-</span> <span class="fu">col</span>(startPointDiffer) <span class="sc">*</span> neg_mask</span>
<span id="cb6-7"><a href="#cb6-7" tabindex="-1"></a>sitesPointsComponet <span class="ot">&lt;-</span> <span class="fu">max.col</span>(col_idx_mat, <span class="at">ties.method =</span> <span class="st">&quot;last&quot;</span>)</span></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="co"># Original: unlist(lapply(...)) per element</span></span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a>sitesPointsComponet_pct <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">lapply</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(...), component_which, ...))</span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a></span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a><span class="co"># Optimized: cbind matrix indexing</span></span>
<span id="cb7-5"><a href="#cb7-5" tabindex="-1"></a>sitesPointsComponet_pct <span class="ot">&lt;-</span> sitesComponet_pct[</span>
<span id="cb7-6"><a href="#cb7-6" tabindex="-1"></a>  <span class="fu">cbind</span>(<span class="fu">seq_along</span>(sitesPointsComponet), sitesPointsComponet)</span>
<span id="cb7-7"><a href="#cb7-7" tabindex="-1"></a>]</span></code></pre></div>
</div>
<div id="fixed-growing-granges-on2-to-on" class="section level3">
<h3>6. Fixed Growing GRanges (O(n^2) to O(n))</h3>
<p><strong>File</strong>: <code>R/makeGuitarTxdb.R</code></p>
<p>The original appended GRanges in a loop, causing quadratic memory
copies:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="co"># Original: O(n^2) - copies entire object each iteration</span></span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a><span class="cf">for</span> (...) {</span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a>  txComponentGRange <span class="ot">&lt;-</span> <span class="fu">c</span>(txComponentGRange, temp_gr)</span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a>}</span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a></span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a><span class="co"># Optimized: O(n) - list accumulation, single concatenation</span></span>
<span id="cb8-7"><a href="#cb8-7" tabindex="-1"></a>gr_list <span class="ot">&lt;-</span> <span class="fu">vector</span>(<span class="st">&quot;list&quot;</span>, expected_length)</span>
<span id="cb8-8"><a href="#cb8-8" tabindex="-1"></a><span class="cf">for</span> (...) {</span>
<span id="cb8-9"><a href="#cb8-9" tabindex="-1"></a>  gr_list[[k]] <span class="ot">&lt;-</span> temp_gr</span>
<span id="cb8-10"><a href="#cb8-10" tabindex="-1"></a>}</span>
<span id="cb8-11"><a href="#cb8-11" tabindex="-1"></a>txComponentGRange <span class="ot">&lt;-</span> <span class="fu">do.call</span>(c, gr_list)</span></code></pre></div>
</div>
<div id="bug-fixes" class="section level3">
<h3>7. Bug Fixes</h3>
<ul>
<li><p><strong>Duplicate rectangle annotations</strong>: Lines 96-101 of
the original <code>GuitarPlot.R</code> were an exact copy of lines
89-94, drawing every transcript component rectangle twice. Removed the
duplicate.</p></li>
<li><p><strong>GuitarTxdb save/load broken</strong>: The original
<code>save()</code> function correctly wrote binary RData, but
<code>load</code> used <code>read.table()</code> to read it back. Fixed
to use <code>load()</code> with an isolated environment.</p></li>
</ul>
</div>
</div>
<div id="numerical-verification" class="section level2">
<h2>Numerical Verification</h2>
<p>For deterministic runs (no CI), GuitarFast produces
<strong>bit-identical</strong> density curves:</p>
<pre><code>Max |y_original - y_optimized| = 0.00e+00</code></pre>
<p>For CI runs, the curves are statistically equivalent within bootstrap
variance. The CI bounds differ only due to random resampling, not due to
algorithmic changes.</p>
</div>
<div id="api-compatibility" class="section level2">
<h2>API Compatibility</h2>
<p>GuitarFast maintains full API compatibility with Guitar v2.26.0. The
only default change:</p>
<table>
<thead>
<tr class="header">
<th align="left">Parameter</th>
<th align="left">Guitar</th>
<th align="left">GuitarFast</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left"><code>CI_ResamplingTime</code></td>
<td align="left">1000</td>
<td align="left">200</td>
</tr>
</tbody>
</table>
<p>All other function signatures, parameter names, and return types are
identical.</p>
</div>
<div id="session-info" class="section level2">
<h2>Session Info</h2>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a><span class="co">#&gt; R version 4.4.3 (2025-02-28)</span></span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a><span class="co">#&gt; Platform: x86_64-conda-linux-gnu</span></span>
<span id="cb10-4"><a href="#cb10-4" tabindex="-1"></a><span class="co">#&gt; Running under: Ubuntu 24.04.4 LTS</span></span>
<span id="cb10-5"><a href="#cb10-5" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb10-6"><a href="#cb10-6" tabindex="-1"></a><span class="co">#&gt; Matrix products: default</span></span>
<span id="cb10-7"><a href="#cb10-7" tabindex="-1"></a><span class="co">#&gt; BLAS/LAPACK: /home/yangli/software/micromamba/envs/guitar/lib/libopenblasp-r0.3.30.so;  LAPACK version 3.12.0</span></span>
<span id="cb10-8"><a href="#cb10-8" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb10-9"><a href="#cb10-9" tabindex="-1"></a><span class="co">#&gt; locale:</span></span>
<span id="cb10-10"><a href="#cb10-10" tabindex="-1"></a><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span id="cb10-11"><a href="#cb10-11" tabindex="-1"></a><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span id="cb10-12"><a href="#cb10-12" tabindex="-1"></a><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span id="cb10-13"><a href="#cb10-13" tabindex="-1"></a><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span id="cb10-14"><a href="#cb10-14" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb10-15"><a href="#cb10-15" tabindex="-1"></a><span class="co">#&gt; time zone: America/Chicago</span></span>
<span id="cb10-16"><a href="#cb10-16" tabindex="-1"></a><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span id="cb10-17"><a href="#cb10-17" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb10-18"><a href="#cb10-18" tabindex="-1"></a><span class="co">#&gt; attached base packages:</span></span>
<span id="cb10-19"><a href="#cb10-19" tabindex="-1"></a><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span id="cb10-20"><a href="#cb10-20" tabindex="-1"></a><span class="co">#&gt; [8] base     </span></span>
<span id="cb10-21"><a href="#cb10-21" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb10-22"><a href="#cb10-22" tabindex="-1"></a><span class="co">#&gt; other attached packages:</span></span>
<span id="cb10-23"><a href="#cb10-23" tabindex="-1"></a><span class="co">#&gt;  [1] GuitarFast_1.0.0       dplyr_1.2.0            knitr_1.51            </span></span>
<span id="cb10-24"><a href="#cb10-24" tabindex="-1"></a><span class="co">#&gt;  [4] ggplot2_4.0.2          magrittr_2.0.4         rtracklayer_1.66.0    </span></span>
<span id="cb10-25"><a href="#cb10-25" tabindex="-1"></a><span class="co">#&gt;  [7] GenomicFeatures_1.58.0 AnnotationDbi_1.68.0   Biobase_2.66.0        </span></span>
<span id="cb10-26"><a href="#cb10-26" tabindex="-1"></a><span class="co">#&gt; [10] GenomicRanges_1.58.0   GenomeInfoDb_1.42.0    IRanges_2.40.0        </span></span>
<span id="cb10-27"><a href="#cb10-27" tabindex="-1"></a><span class="co">#&gt; [13] S4Vectors_0.44.0       BiocGenerics_0.52.0   </span></span>
<span id="cb10-28"><a href="#cb10-28" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb10-29"><a href="#cb10-29" tabindex="-1"></a><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span id="cb10-30"><a href="#cb10-30" tabindex="-1"></a><span class="co">#&gt;  [1] KEGGREST_1.46.0             SummarizedExperiment_1.36.0</span></span>
<span id="cb10-31"><a href="#cb10-31" tabindex="-1"></a><span class="co">#&gt;  [3] gtable_0.3.6                rjson_0.2.23               </span></span>
<span id="cb10-32"><a href="#cb10-32" tabindex="-1"></a><span class="co">#&gt;  [5] xfun_0.56                   bslib_0.10.0               </span></span>
<span id="cb10-33"><a href="#cb10-33" tabindex="-1"></a><span class="co">#&gt;  [7] lattice_0.22-9              generics_0.1.4             </span></span>
<span id="cb10-34"><a href="#cb10-34" tabindex="-1"></a><span class="co">#&gt;  [9] vctrs_0.7.1                 tools_4.4.3                </span></span>
<span id="cb10-35"><a href="#cb10-35" tabindex="-1"></a><span class="co">#&gt; [11] bitops_1.0-9                curl_7.0.0                 </span></span>
<span id="cb10-36"><a href="#cb10-36" tabindex="-1"></a><span class="co">#&gt; [13] parallel_4.4.3              tibble_3.3.1               </span></span>
<span id="cb10-37"><a href="#cb10-37" tabindex="-1"></a><span class="co">#&gt; [15] RSQLite_2.4.6               blob_1.3.0                 </span></span>
<span id="cb10-38"><a href="#cb10-38" tabindex="-1"></a><span class="co">#&gt; [17] pkgconfig_2.0.3             Matrix_1.7-4               </span></span>
<span id="cb10-39"><a href="#cb10-39" tabindex="-1"></a><span class="co">#&gt; [19] RColorBrewer_1.1-3          S7_0.2.1                   </span></span>
<span id="cb10-40"><a href="#cb10-40" tabindex="-1"></a><span class="co">#&gt; [21] lifecycle_1.0.5             GenomeInfoDbData_1.2.13    </span></span>
<span id="cb10-41"><a href="#cb10-41" tabindex="-1"></a><span class="co">#&gt; [23] farver_2.1.2                compiler_4.4.3             </span></span>
<span id="cb10-42"><a href="#cb10-42" tabindex="-1"></a><span class="co">#&gt; [25] Rsamtools_2.22.0            Biostrings_2.74.0          </span></span>
<span id="cb10-43"><a href="#cb10-43" tabindex="-1"></a><span class="co">#&gt; [27] codetools_0.2-20            htmltools_0.5.9            </span></span>
<span id="cb10-44"><a href="#cb10-44" tabindex="-1"></a><span class="co">#&gt; [29] sass_0.4.10                 RCurl_1.98-1.17            </span></span>
<span id="cb10-45"><a href="#cb10-45" tabindex="-1"></a><span class="co">#&gt; [31] yaml_2.3.12                 pillar_1.11.1              </span></span>
<span id="cb10-46"><a href="#cb10-46" tabindex="-1"></a><span class="co">#&gt; [33] crayon_1.5.3                jquerylib_0.1.4            </span></span>
<span id="cb10-47"><a href="#cb10-47" tabindex="-1"></a><span class="co">#&gt; [35] BiocParallel_1.40.0         cachem_1.1.0               </span></span>
<span id="cb10-48"><a href="#cb10-48" tabindex="-1"></a><span class="co">#&gt; [37] DelayedArray_0.32.0         abind_1.4-8                </span></span>
<span id="cb10-49"><a href="#cb10-49" tabindex="-1"></a><span class="co">#&gt; [39] tidyselect_1.2.1            digest_0.6.39              </span></span>
<span id="cb10-50"><a href="#cb10-50" tabindex="-1"></a><span class="co">#&gt; [41] restfulr_0.0.16             fastmap_1.2.0              </span></span>
<span id="cb10-51"><a href="#cb10-51" tabindex="-1"></a><span class="co">#&gt; [43] grid_4.4.3                  cli_3.6.5                  </span></span>
<span id="cb10-52"><a href="#cb10-52" tabindex="-1"></a><span class="co">#&gt; [45] SparseArray_1.6.0           S4Arrays_1.6.0             </span></span>
<span id="cb10-53"><a href="#cb10-53" tabindex="-1"></a><span class="co">#&gt; [47] XML_3.99-0.22               withr_3.0.2                </span></span>
<span id="cb10-54"><a href="#cb10-54" tabindex="-1"></a><span class="co">#&gt; [49] scales_1.4.0                UCSC.utils_1.2.0           </span></span>
<span id="cb10-55"><a href="#cb10-55" tabindex="-1"></a><span class="co">#&gt; [51] bit64_4.6.0-1               rmarkdown_2.30             </span></span>
<span id="cb10-56"><a href="#cb10-56" tabindex="-1"></a><span class="co">#&gt; [53] XVector_0.46.0              httr_1.4.8                 </span></span>
<span id="cb10-57"><a href="#cb10-57" tabindex="-1"></a><span class="co">#&gt; [55] matrixStats_1.5.0           bit_4.6.0                  </span></span>
<span id="cb10-58"><a href="#cb10-58" tabindex="-1"></a><span class="co">#&gt; [57] png_0.1-8                   memoise_2.0.1              </span></span>
<span id="cb10-59"><a href="#cb10-59" tabindex="-1"></a><span class="co">#&gt; [59] evaluate_1.0.5              BiocIO_1.16.0              </span></span>
<span id="cb10-60"><a href="#cb10-60" tabindex="-1"></a><span class="co">#&gt; [61] rlang_1.1.7                 glue_1.8.0                 </span></span>
<span id="cb10-61"><a href="#cb10-61" tabindex="-1"></a><span class="co">#&gt; [63] DBI_1.2.3                   jsonlite_2.0.0             </span></span>
<span id="cb10-62"><a href="#cb10-62" tabindex="-1"></a><span class="co">#&gt; [65] R6_2.6.1                    MatrixGenerics_1.18.0      </span></span>
<span id="cb10-63"><a href="#cb10-63" tabindex="-1"></a><span class="co">#&gt; [67] GenomicAlignments_1.42.0    zlibbioc_1.52.0</span></span></code></pre></div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
